Tests/BuildTests/IncrementalBuildTests.swift

```diff
@@ -15,7 +15,130 @@ import Basic
+    }
+}
+
+typealias BuildLog = [Builds : BuildResult]
 
+// Seems like this is too tied to the build's log format
+struct BuildResult : Equatable {
+    static func == (lhs: BuildResult, rhs: BuildResult) -> Bool {
+        return
+            lhs.path == rhs.path &&
+            lhs.logWasNil == rhs.logWasNil &&
+            lhs.isNullBuild == rhs.isNullBuild &&
+            lhs.modules.count == rhs.modules.count &&
+            zip(lhs.modules, rhs.modules).reduce(true) { $0 && $1.0.fileCount == $1.1.fileCount && $1.0.name == $1.1.name } &&
+            lhs.linkedBinaries.count == rhs.linkedBinaries.count &&
+            zip(lhs.linkedBinaries, rhs.linkedBinaries).reduce(true) { $0 && $1.0.path == $1.1.path && $1.0.name == $1.1.name } &&
+            lhs.buildProducts == rhs.buildProducts
+    }
+
+    let path: AbsolutePath
+    let logWasNil: Bool
+    let isNullBuild: Bool
+    var isNotNullBuild: Bool { return !isNullBuild }
+    let modules: [(name: String, fileCount: Int)]
+    let linkedBinaries: [(path: String, name: String)]
+    var totalFileCount: Int { return modules.reduce(0) { $0 + $1.fileCount } }
+    /// An array of the names/paths of files in ./build/debug/ right after a build
+    let buildProducts: [String]
+    init(_ log: String?, path: AbsolutePath) {
+        self.path = path
+        if let log = log {
+            logWasNil = false
+            isNullBuild = log == ""
+            let logLines = log.components(separatedBy: "\n")
+            let compileLines = logLines.filter {$0.hasPrefix("Compile Swift Module") && $0.contains("(") && $0.hasSuffix(" sources)") }
+            modules = compileLines
+                .map { ($0.components(separatedBy: "\'")[1], Int($0.components(separatedBy: "(").last!.components(separatedBy: CharacterSet.whitespaces)[0])!) }
+            linkedBinaries = logLines
+                .filter { $0.hasPrefix("Linking ") && $0.contains("/") }
+                .map {
+                    let path = $0.substring(from: $0.range(of: " ")!.upperBound)
+                    let name = path.components(separatedBy: "/").last!
+                    return (path, name)
+            }
+            
+            // So the idea here is to log everything in ".builds/debug/", except
+            // for "ModuleCache", because that doesn't seem relevant. We crash
+            // without the `&& !$0.hasSuffix(".dSYM")` part, probably because
+            // localFileSystem.isDirectory() seems to return true for .dSYM files?
+            var bp = try! localFileSystem.getDirectoryContents(path.appending(components: ".build", "debug"))
+                .filter { $0 != "ModuleCache" }
+            // Get the list of "files" that are actually directories...
+            var bpd = bp.filter { localFileSystem.isDirectory(path.appending(components: ".build", "debug", $0)) && !$0.hasSuffix(".dSYM") }
```

Comment:
I believe it should be possible to simplify this by using Basic.walk (see `PathShims.swift`)...